mobile/android/android-device.ts

```diff
@@ -185,9 +188,166 @@ export class AndroidDevice implements Mobile.IDevice {
+        }).future<void>()();
+    }
+
+    private openDebuggerClient(url: string): IFuture<void> {
+        return (() => {
+            this.$opener.open(url, "chrome");
+        }).future<void>()();
+    }
+
+    private printDebugPort(packageName: string): void {
+        var res = this.$childProcess.spawnFromEvent(this.adb, ["shell", "am", "broadcast", "-a", packageName + "-GetDgbPort"], "exit").wait();
+        this.$logger.info(res.stdout);
+    }
+
+    private attachDebugger(packageName: string): void {
+        var startDebuggerCommand = this.composeCommand("shell am broadcast -a \"%s-Debug\" --ez enable true", packageName);
+        var port = options["debug-port"];
+        if (port > 0) {
+            startDebuggerCommand += " --ei debuggerPort " + options["debug-port"];
+            this.$childProcess.exec(startDebuggerCommand).wait();
+        } else {
+            var res = this.$childProcess.spawnFromEvent(this.adb, ["shell", "am", "broadcast", "-a", packageName + "-Debug", "--ez", "enable", "true"], "exit").wait();
+            var match = res.stdout.match(/result=(\d)+/);
+            if (match) {
+                port = match[0].substring(7);
+            } else {
+                port = 0;
+            }
+        }
+        if (port > 0) {
+            this.tcpForward(port, port).wait();
+        }
+        this.startDebuggerClient(port).wait();
+        this.openDebuggerClient(this.defaultNodeInspectorUrl + "?port=" + port).wait();
+    }
+
+    private detachDebugger(packageName: string): void {
+        var stopDebuggerCommand = this.composeCommand("shell am broadcast -a \"%s-Debug\" --ez enable false", packageName);
+        this.$childProcess.exec(stopDebuggerCommand).wait();
+    }
+
+    private startAppWithDebugger(packageFile: string, packageName: string): void {
+        var uninstallCommand = this.composeCommand("shell pm uninstall \"%s\"", packageName)
+			    this.$childProcess.exec(uninstallCommand).wait();
+
+        var installCommand = this.composeCommand("install -r \"%s\"", packageFile);
+        this.$childProcess.exec(installCommand).wait();
+
+        var port = options["debug-port"];
+
+        var clearDebugEnvironmentCommand = this.composeCommand('shell rm "/sdcard/Android/data/%s/files/envDebug.out"', packageName);
+        this.$childProcess.exec(clearDebugEnvironmentCommand).wait();
+
+        this.startPackageOnDevice(packageName).wait();
+
+        var dbgPort = this.startAndGetPort(packageName).wait();
+
+        if (dbgPort > 0) {
+            this.tcpForward(dbgPort, dbgPort).wait();
+            this.startDebuggerClient(dbgPort).wait();
+            this.openDebuggerClient(this.defaultNodeInspectorUrl + "?port=" + dbgPort).wait();
+        }
+    }
+
+    public debug(packageFile: string, packageName: string): IFuture<void> {
+        return (() => {
+            if (options["get-port"]) {
```

Comment:
I believe that it is much better to nuke this function and instead implement compound commands.